### ISSUE TITLE
Unfair comparison with CoTTA in segmentation tasks with lower learning rate

### ISSUE THREAD

**Issue by hahahatatata** [Yangsenqiao/vida#4]
Hi, I want to first congratulate you on your excellent work!

However, I have experimented on CoTTA with a learning rate of 3e-4 and its performance is amazing enough even without test-time augmentation:
Revisiting 0: Fog = 70.17	Night = 44.49	Rain=65.23	Snow = 62.74

Comparing with yours in Table 4:
Revisiting 0: Fog = 71.6	Night = 43.2 Rain = 66.0 Snow = 63.4

It seems that not only in this paper but also in VDP and SVDP, your segmentation results are not fair to baselines with a learning rate of 7.5e-6. Kindly hope you can double check it. 

Regards.

**Comment by liujiaming1996** [comment]
> Hi, I want to first congratulate you on your excellent work!
> 
> However, I have experimented on CoTTA with a learning rate of 3e-4 and its performance is amazing enough even without test-time augmentation: Revisiting 0: Fog = 70.17 Night = 44.49 Rain=65.23 Snow = 62.74
> 
> Comparing with yours in Table 4: Revisiting 0: Fog = 71.6 Night = 43.2 Rain = 66.0 Snow = 63.4
> 
> It seems that not only in this paper but also in VDP and SVDP, your segmentation results are not fair to baselines with a learning rate of 7.5e-6. Kindly hope you can double check it.
> 
> Regards.

Dear hahahatatata, 

Firstly, thank you for your comments, and we appreciate your recognition of our paper.

The segmentation experimental results for CoTTA are copied from the original CoTTA paper (https://openaccess.thecvf.com/content/CVPR2022/papers/Wang_Continual_Test-Time_Domain_Adaptation_CVPR_2022_paper.pdf), including the tables in both the main body and the appendix.

We haven't attempted to reproduce the CoTTA experiments under our segmentation code framework. We will explore the experimental setting you mentioned as soon as possible.

Simultaneously, regarding the use of test-time augmentation and its efficacy for CTTA tasks, we are currently investigating this aspect. Based on our current experimental results, introducing augmentation appears to enhance the model's/adapter‘s stability over longer rounds of experiments.

If you are also interested in dense prediction CTTA tasks, we hope to engage in further discussion with you.

Regards

**Comment by hahahatatata** [comment]
> Simultaneously

Hi Jiaming,

Thanks for your reply. Yes, I am also doing the dense CTTA task.

You can simply change the learning rate in CoTTA's codebase and reproduce the experiments. The main concern is that your results and visualizations may not correctly illustrate your contributions without using the same learning rate and future researchers may also follow your steps in their papers by copying your results. 

In my experiments, the test-time augmentation and stochastic restoration can be effective with higher learning rates. Therefore, I believe your learning rate is optimal for both CoTTA and your models. Also, due to the stochastic restoration, the performance of CoTTA is not stable and it can sometimes have even better results than yours. Therefore, I also suggest performing your experiments again without stochastic restoration. BTW, do check the order of your data streaming because CoTTA does not have such a mechanism (like sorting) but I find it somewhat sensitive.

Regards

**Comment by liujiaming1996** [comment]
> > Simultaneously
> 
> Hi Jiaming,
> 
> Thanks for your reply. Yes, I am also doing the dense CTTA task. I believe you are already one of the leading researchers in this task so I want to point out this problem in this issue.
> 
> You can simply change the learning rate in CoTTA's codebase and reproduce the experiments. The main concern is that your results and visualizations may not correctly illustrate your contributions without using the same learning rate and future researchers may also follow your steps in their papers by copying your results.
> 
> In my experiments, the test-time augmentation and stochastic restoration can be effective with higher learning rates. Therefore, I believe your learning rate is optimal for both CoTTA and your models. Also, due to the stochastic restoration, the performance of CoTTA is not stable and it can sometimes have even better results than yours. Therefore, I also suggest performing your experiments again without stochastic restoration. BTW, do check the order of your data streaming because CoTTA does not have such a mechanism (like sorting) but I find it somewhat sensitive.
> 
> Regards

Dear hahahatatata,
    Thank you for pointing out the issue. I will conduct new additinal experiments by adjusting the learning rates as suggested in the original CoTTA paper. I will also discuss the impact of learning rates on segmentation CTTA in the paper. However, I'm curious if you have run experiments with high learning rates over 10 rounds. I suspect that using a high learning rate for a model-based CTTA method may lead to catastrophic forgetting in the long-term adaptation process. Anyway, we will conduct more experiments and incorporate meaningful results into the paper.

    Meanwhile, regarding the statement, 'Therefore, I also suggest performing your experiments again without stochastic restoration,' I'm not entirely clear. If it pertains to reproducing the CoTTA code, I still hope to preserve all contributions made by the original authors. In summary, thank you for the constructive suggestions. We will conduct supplementary experiments on the original CoTTA paper's codebase, the shift dataset challenge codebase, and our own codebase to investigate the impact of LR on segmentation CTTA.
    
   Regards

**Comment by liujiaming1996** [comment]
> check the order of your data streaming because CoTTA

Thank you for your detailed suggestion“check the order of your data streaming because CoTTA does not have such a mechanism (like sorting) but I find it somewhat sensitive.“  I will conduct further experiments on the data streaming order as well.

**Comment by liujiaming1996** [comment]
> ue to the stochastic restoration, the performance of CoTTA is not stable and it can sometimes have even better results than yours.

I understand your point, and I will control the random seed in the experiments. We'll run multiple sets of experiments and average the performance for more robust results.

**Comment by hahahatatata** [comment]
> > > Simultaneously
> > 
> > 
> > Hi Jiaming,
> > Thanks for your reply. Yes, I am also doing the dense CTTA task. I believe you are already one of the leading researchers in this task so I want to point out this problem in this issue.
> > You can simply change the learning rate in CoTTA's codebase and reproduce the experiments. The main concern is that your results and visualizations may not correctly illustrate your contributions without using the same learning rate and future researchers may also follow your steps in their papers by copying your results.
> > In my experiments, the test-time augmentation and stochastic restoration can be effective with higher learning rates. Therefore, I believe your learning rate is optimal for both CoTTA and your models. Also, due to the stochastic restoration, the performance of CoTTA is not stable and it can sometimes have even better results than yours. Therefore, I also suggest performing your experiments again without stochastic restoration. BTW, do check the order of your data streaming because CoTTA does not have such a mechanism (like sorting) but I find it somewhat sensitive.
> > Regards
> 
> Dear hahahatatata, Thank you for pointing out the issue. I will conduct new additinal experiments by adjusting the learning rates as suggested in the original CoTTA paper. I will also discuss the impact of learning rates on segmentation CTTA in the paper. However, I'm curious if you have run experiments with high learning rates over 10 rounds. I suspect that using a high learning rate for a model-based CTTA method may lead to catastrophic forgetting in the long-term adaptation process. Anyway, we will conduct more experiments and incorporate meaningful results into the paper.
> 
> Meanwhile, regarding the statement, 'Therefore, I also suggest performing your experiments again without stochastic restoration,' I'm not entirely clear. If it pertains to reproducing the CoTTA code, I still hope to preserve all contributions made by the original authors. In summary, thank you for the constructive suggestions. We will conduct supplementary experiments on the original CoTTA paper's codebase, the shift dataset challenge codebase, and our own codebase to investigate the impact of LR on segmentation CTTA.
> 
> Regards
Hi Jiaming, 

Yes, my experiments show that a high learning rate may lead to an earlier catastrophic forgetting. You can visualize the output of the student network and see the degradation. Lower learning rate and EMA are similar to the weights regularizations which slows down the optimization as well as the catastrophic forgetting but is not ideal for short-term predictions. Note that ACDC can see the visited data after the first rounds and thus the catastrophic forgetting is not obvious (similar to multi-epoch training).

For the statement 'Therefore, I also suggest performing your experiments again without stochastic restoration', I find during my experiments that stochastic restoration can cause instability due to random sampling. Sometimes bad weights are restored while sometimes good weights are restored. Therefore, removing this part can at least ensure reproducibility and make your experiments even more convincible. 

Hope these suggestions can help you and make your experiments even better.

**Comment by liujiaming1996** [comment]
> Hope these suggestions can help you and make your experiments even better.

Thank you for your suggestion. We find your insights very valuable. Would it be convenient to have further discussion with you? My WeChat ID is q764316889, and my email is jiamingliu@stu.edu.pku.cn.

**Comment by hahahatatata** [comment]
Hi, I have reproduced the CoTTA results with a learning rate of 3e-4 and the full list of test-time augmentation. Hope this helps.

writing results to work_dirs/res.pkl
per class results:

```
+---------------+-------+-------+
| Class         | IoU   | Acc   |
+---------------+-------+-------+
| road          | 93.91 | 99.39 |
| sidewalk      | 63.29 | 71.37 |
| building      | 80.92 | 97.2  |
| wall          | 56.04 | 67.8  |
| fence         | 24.8  | 26.22 |
| pole          | 51.08 | 59.08 |
| traffic light | 44.12 | 48.93 |
| traffic sign  | 73.45 | 90.54 |
| vegetation    | 87.4  | 91.82 |
| terrain       | 71.18 | 88.68 |
| sky           | 97.86 | 99.09 |
| person        | 68.36 | 82.22 |
| rider         | 71.61 | 88.57 |
| car           | 87.34 | 95.18 |
| truck         | 77.82 | 82.12 |
| bus           | 93.71 | 97.49 |
| train         | 89.06 | 96.59 |
| motorcycle    | 59.25 | 66.72 |
| bicycle       | 66.83 | 86.9  |
+---------------+-------+-------+
Summary:

+--------+-------+-------+-------+
| Scope  | mIoU  | mAcc  | aAcc  |
+--------+-------+-------+-------+
| global | 71.48 | 80.84 | 93.92 |
+--------+-------+-------+-------+
``` 

```
+---------------+-------+-------+
| Class         | IoU   | Acc   |
+---------------+-------+-------+
| road          | 88.6  | 98.94 |
| sidewalk      | 51.0  | 56.13 |
| building      | 67.16 | 89.11 |
| wall          | 34.2  | 44.25 |
| fence         | 23.56 | 25.83 |
| pole          | 50.41 | 61.76 |
| traffic light | 47.59 | 61.86 |
| traffic sign  | 47.27 | 52.6  |
| vegetation    | 41.33 | 85.55 |
| terrain       | 26.1  | 65.35 |
| sky           | 0.96  | 0.96  |
| person        | 52.96 | 76.38 |
| rider         | 44.73 | 55.25 |
| car           | 77.65 | 91.91 |
| truck         | 18.12 | 71.32 |
| bus           | 43.21 | 44.75 |
| train         | 58.79 | 62.21 |
| motorcycle    | 31.6  | 35.38 |
| bicycle       | 39.02 | 46.29 |
+---------------+-------+-------+
Summary:

+--------+-------+-------+------+
| Scope  | mIoU  | mAcc  | aAcc |
+--------+-------+-------+------+
| global | 44.44 | 59.25 | 72.4 |
+--------+-------+-------+------+
```

```
+---------------+-------+-------+
| Class         | IoU   | Acc   |
+---------------+-------+-------+
| road          | 85.13 | 98.15 |
| sidewalk      | 52.56 | 56.83 |
| building      | 90.81 | 97.17 |
| wall          | 45.2  | 59.55 |
| fence         | 32.44 | 36.85 |
| pole          | 60.85 | 72.62 |
| traffic light | 70.38 | 77.12 |
| traffic sign  | 69.96 | 82.33 |
| vegetation    | 90.96 | 94.92 |
| terrain       | 59.37 | 84.36 |
| sky           | 98.09 | 98.58 |
| person        | 56.6  | 69.9  |
| rider         | 61.47 | 80.69 |
| car           | 82.79 | 91.24 |
| truck         | 36.79 | 61.74 |
| bus           | 76.0  | 79.39 |
| train         | 72.49 | 87.07 |
| motorcycle    | 51.61 | 61.74 |
| bicycle       | 60.71 | 77.92 |
+---------------+-------+-------+
Summary:

+--------+-------+-------+-------+
| Scope  | mIoU  | mAcc  | aAcc  |
+--------+-------+-------+-------++---------------+-------+-------+
| Class         | IoU   | Acc   |
+---------------+-------+-------+
| road          | 82.66 | 98.57 |
| sidewalk      | 46.8  | 56.91 |
| building      | 86.83 | 97.84 |
| wall          | 51.16 | 62.23 |
| fence         | 45.15 | 47.54 |
| pole          | 63.32 | 72.32 |
| traffic light | 76.16 | 82.89 |
| traffic sign  | 72.15 | 82.22 |
| vegetation    | 82.6  | 85.93 |
| terrain       | 4.72  | 5.33  |
| sky           | 97.79 | 98.74 |
| person        | 68.69 | 83.46 |
| rider         | 47.43 | 80.18 |
| car           | 85.81 | 94.54 |
| truck         | 51.2  | 60.09 |
| bus           | 60.04 | 72.18 |
| train         | 86.14 | 88.23 |
| motorcycle    | 25.18 | 27.87 |
| bicycle       | 55.95 | 74.22 |
+---------------+-------+-------+
Summary:

+--------+-------+-------+-------+
| Scope  | mIoU  | mAcc  | aAcc  |
+--------+-------+-------+-------+
| global | 62.62 | 72.17 | 90.93 |
+--------+-------+-------+-------+
| global | 66.01 | 77.27 | 92.83 |
+--------+-------+-------+-------+
```

```
+---------------+-------+-------+
| Class         | IoU   | Acc   |
+---------------+-------+-------+
| road          | 82.66 | 98.57 |
| sidewalk      | 46.8  | 56.91 |
| building      | 86.83 | 97.84 |
| wall          | 51.16 | 62.23 |
| fence         | 45.15 | 47.54 |
| pole          | 63.32 | 72.32 |
| traffic light | 76.16 | 82.89 |
| traffic sign  | 72.15 | 82.22 |
| vegetation    | 82.6  | 85.93 |
| terrain       | 4.72  | 5.33  |
| sky           | 97.79 | 98.74 |
| person        | 68.69 | 83.46 |
| rider         | 47.43 | 80.18 |
| car           | 85.81 | 94.54 |
| truck         | 51.2  | 60.09 |
| bus           | 60.04 | 72.18 |
| train         | 86.14 | 88.23 |
| motorcycle    | 25.18 | 27.87 |
| bicycle       | 55.95 | 74.22 |
+---------------+-------+-------+
Summary:

+--------+-------+-------+-------+
| Scope  | mIoU  | mAcc  | aAcc  |
+--------+-------+-------+-------+
| global | 62.62 | 72.17 | 90.93 |
+--------+-------+-------+-------+
```
